why-choice/index.qmd

```diff
@@ -0,0 +1,265 @@
+-   Pydantic is widely used, extensively tested, and has good
+    documentation.
+:::
+
+::: column
+#### Drawbacks
+
+-   We would need to translate the Data Package standard into Pydantic
+    models to be able to use Pydantic for checking objects. However, we
+    could use `datamodel_code_generator` to automate this process (see
+    on the left).
+-   While integrating our existing metadata classes with Pydantic models
+    would lead to a single representation of the standard in Sprout, our
+    previous experience with generating classes from the schema has
+    shown that the resulting class system is more complex than what we
+    would like. If we choose Pydantic, a similar exercise of simplifying
+    the class system without losing compatibility with the standard is
+    foreseen. This would lead to losing the ability to receive updates
+    to the standard automatically.
+-   As Pydantic models have a `schema` attribute, we would need to alias
+    the `schema` attribute of our resource metadata class internally to
+    avoid ambiguity.
+-   Currently, our metadata classes are allowed to represent partial
+    metadata objects (e.g. just one field specified). Adding checks
+    directly in these classes at instantiation time could lead to some
+    difficulties with instantiating partial metadata objects, which
+    might not meet the standard in that form.
+:::
+:::::
+
+### jsonschema
+
+[jsonschema](https://python-jsonschema.readthedocs.io/en/stable/) is the
+most widely used implementation of the [JSON Schema
+specification](https://json-schema.org/) in Python. Its main use case is
+checking that a JSON object conforms to a JSON schema.
+
+::::: columns
+::: column
+#### Benefits
+
+-   Can check JSON metadata files against the JSON schema of the Data
+    Package standard directly.
+-   Matches the schema exactly, e.g., it disallows empty lists.
+-   Does not introduce a parallel set of classes representing the
+    standard.
+-   We can receive updates to the standard automatically by downloading
+    the new version of the schema.
+-   It is possible to collect all errors in a single round of checks.
+-   It is possible to filter errors both by source (which field did not
+    match) and by cause (why that field did not match).
+-   `jsonschema` is widely used, extensively tested, and has good
+    documentation.
+:::
+
+::: column
+#### Drawbacks
+
+-   Does not include optional recommendations in the textual description
+    of the standard.
+-   Cannot check package, resource and table metadata separately out of
+    the box, but this would be fairly easy to implement.
+-   Does not provide error codes as such, but it does expose the cause
+    of the error (e.g. a missing required field, a mismatched pattern,
+    etc.) in a property other than the error message as well.
+:::
+:::::
+
+## Decision outcome
+
+::: content-hidden
+What decision was made, use the form "We decided on CHOICE because of
+REASONS."
+:::
+
+TBD
```

Comment:
Nice! Which tool are you leaning towards? I sense you think `jsonschema` is the tool you'd choose?

Reply:
I'm also feeling like that's the one I would choose, basically because it's the simplest choice and does what it says on the tin. My only concern is that for the same reasons it's also the most restrictive. Like, we could make Pydantic do pretty much anything, but `jsonschema` will always just compare a dictionary against a JSON schema. Plus, if we ever want to add similar checks for objects that are not based on a JSON schema, these will likely use Pydantic, and then we may end up using both? But maybe we shouldn't worry about that at this stage.

Reply:
Great! Let's use that then :grin:

Reply:
@martonvago Are you thinking of any particular objects that would not be based on a JSON schema? I would think that all checks would be against the Data Package standard, i.e., a JSON schema?

Reply:
Okay, I will add the conclusion then if there are no objections 😊